kibana-reports/public/components/context_menu/context_menu.js

```diff
@@ -0,0 +1,251 @@
+    const popoverScreen = document.querySelectorAll("#kibana-app");
+    if (popoverScreen) {
+      try {
+        const reportPopover = document.createElement("div");
+        reportPopover.innerHTML = popoverMenu();
+        popoverScreen[0].appendChild(reportPopover.children[0]);
+        $('#reportPopover').show();
+      } catch (e) {
+        console.log("error displaying menu:", e);
+      } 
+    }
+  })
+
+  // generate PDF onclick
+  $(document).on('click', '#generatePDF', function() {
+    let timeRanges = getTimeFieldsFromUrl();
+    const queryUrl = replaceQueryURL();
+    generateInContextReport(timeRanges, queryUrl, 'pdf');
+  })
+
+  // generate PNG onclick
+  $(document).on('click', '#generatePNG', function() {
+    let timeRanges = getTimeFieldsFromUrl();
+    const queryUrl = replaceQueryURL();
+    generateInContextReport(timeRanges, queryUrl, 'png');
+  })
+
+  // navigate to Create report definition page with report source and pre-set time range
+  $(document).on('click', '#createReportDefinition', function() {
+    contextMenuCreateReportDefinition(this.baseURI);
+  })
+
+  // close popover menu on click outside
+  $('body').on('click', function(e) {
+    if ($(e.target).data('toggle') !== '#downloadReport') {
+      $('#reportPopover').remove();
+    }
+  })
+
+  // close modal/toast 
+  $(document).ready(function() {
+    // close modal with 'x' in upper-right modal
+    $(document).on('click', '#closeReportGenerationModal', function() {
+      $('#reportGenerationProgressModal').remove();
+    }); 
+
+    // close modal with the close EuiButton
+    $(document).on('click', '#closeReportGenerationModalButton', function() {
+      $('#reportGenerationProgressModal').remove();
+    })
+
+    // close the toast that appears upon successful report generation
+    $(document).on('click', '#closeReportSuccessToast', function() {
+      $('#reportSuccessToast').remove();
+    })
+
+    // close the toast that apepars upon failure of report generation
+    $(document).on('click', '#closeReportFailureToast', function() {
+      $('#reportFailureToast').remove();
+    })
+  });
+  
+   const observer = new MutationObserver(function (mutations) {
+ 
+    const navMenu = document.querySelectorAll("span.kbnTopNavMenu__wrapper > div.euiFlexGroup");
+    if(navMenu && navMenu.length && navMenu[0].children.length > 1) {
+      try {
+        const menuItem = document.createElement("div");
+        menuItem.innerHTML = getMenuItem("Reporting");
+        navMenu[0].appendChild(menuItem.children[0])
+        
+      } catch (e) {
+        console.log(e);
+      } finally {
+        observer.disconnect();
+      }
+    }
+   });
+ 
+   // Start observing
+   observer.observe(document.body, { //document.body is node target to observe
+     childList: true, //This is a must have for the observer with subtree
+     subtree: true //Set to true if changes must also be observed in descendants.
+   });
```

Comment:
since we use onhashchange event, properly we don't need to render in onready event

Reply:
Sorry not sure I understand, what's the change to make?

Reply:
for below menu rendering code
```
   const observer = new MutationObserver(function (mutations) {

    const navMenu = document.querySelectorAll("span.kbnTopNavMenu__wrapper > div.euiFlexGroup");
    if(navMenu && navMenu.length && navMenu[0].children.length > 1) {
      try {
        const menuItem = document.createElement("div");
        menuItem.innerHTML = getMenuItem("Reporting");
        navMenu[0].appendChild(menuItem.children[0])

      } catch (e) {
        console.log(e);
      } finally {
        observer.disconnect();
      }
    }
   });

   // Start observing
   observer.observe(document.body, { //document.body is node target to observe
     childList: true, //This is a must have for the observer with subtree
     subtree: true //Set to true if changes must also be observed in descendants.
   });
```



we could remove the render menu code from ```$(document).ready``` since  ```function locationHashChanged()``` already cover the rendering logic
```
$(document).ready(function() {
 ...
}
```

Reply:
Got it- removed code and called `locationHashChanged()` instead